README.md
# focusnode_widgets

Reusable focus node widgets. Can be controlled with 
keyboard or Remote Control (e.g. on Android TV).

## Usage
To use this plugin, add `focusnode_widgets` as a [dependency in your pubspec.yaml file](https://flutter.dev/platform-plugins/).

### Example

``` dart
import 'package:flutter/material.dart';
import 'package:focusnode_widgets/focusnode_widgets.dart';

void main() => runApp(MenuForAndroidTV());

class MenuForAndroidTV extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _MenuForAndroidTV();
  }
}

class _MenuForAndroidTV extends State<MenuForAndroidTV> {
  final String _title = 'focusnode_widgets example for Android TV';
  final String _instructions =
      'Use DPAD up down arrow keyss to navigate and DPAD center key to open the URL';

  bool checked = false;

  @override
  Widget build(BuildContext context) {
    final TextTheme textTheme = Theme.of(context).textTheme;

    void launchUrl(BuildContext context) {
      // put code to launch the URL here
      // on android TV url needs to be launched in WebView as per guidelines,
      // browser can't be used
    }

    HyperLinkMenuItem flutterLink = HyperLinkMenuItem(
      displayText: 'Flutter - Beautiful native apps in record time',
      launchUrl: launchUrl,
      autoFocus: true,
    );

    void _onChanged(BuildContext context, bool selected) {
      checked = selected;
      setState(() {});
    }

    CheckboxListTileMenuItem checkboxListTileMenuItem =
        CheckboxListTileMenuItem(
      title: DefaultTextStyle(
          style: textTheme.headline,
          child: Text('I accept Terms and Conditions')),
      enterTapActionCallback: _onChanged,
      value: checked,
      controlAffinity: ListTileControlAffinity.leading,
    );

    Text nonFocusableText2 = Text('Non focusable text item 2');

    void _handleEnterTapAction(BuildContext context) {
      final snackBar = SnackBar(
        content: Text('Yay! A SnackBar!'),
        action: SnackBarAction(
          label: 'Undo',
          onPressed: () {
            // Some code to undo the change.
          },
        ),
      );
      Scaffold.of(context).showSnackBar(snackBar);
    }

    FocusableSimpleMenuItem focusableSimpleMenuItem = FocusableSimpleMenuItem(
      child: Text('Click, Tap or Press Enter/ OK to show SnackBar'),
      enterTapActionCallback: _handleEnterTapAction,
    );

    Text instructions = Text(_instructions);

    VerticalMenuForAndroidTV verticalMenuForAndroidTV =
        VerticalMenuForAndroidTV(
      menuItems: [
//        keyPrintMenuItem,
//        keyLogShort,
        flutterLink,
        nonFocusableText2,
        checkboxListTileMenuItem,
        focusableSimpleMenuItem,
        instructions,
      ],
      focusedBackgroundDecoration: BoxDecoration(
        border: Border.all(color: Colors.amber[900], width: 2),
      ),
      alignment: Alignment.center,
      constraints: BoxConstraints(
        maxHeight: 40,
        maxWidth: 700,
        minHeight: 10,
        minWidth: 100,
      ),
    );

    DefaultTextStyle menu = DefaultTextStyle(
      style: textTheme.headline,
      child: verticalMenuForAndroidTV,
    );

    return MaterialApp(
      title: _title,
      home: Scaffold(
        appBar: AppBar(title: Text(_title)),
        body: menu,
      ),
    );
  }
}

```


CHANGELOG.md
## [0.0.9] - 4 Sep 2019

* Modified HyperLinkMenuItem to not to call the launchUrl internally, provided callback for the same,
on android TV url needs to be launched in WebView as per guidelines,

## [0.0.8] - 2 Sep 2019

* Bug fix : D-PAD - center (logical select key was not handled)

## [0.0.7] - 2 Sep 2019

* Support of KeyPrintMenuItem

## [0.0.6] - 1 Sep 2019

* It should be possible to accept any Widget in VerticalMenuForAndroidTV


## [0.0.5] - 1 Sep 2019

* Support of CheckboxListTileMenuItem


## [0.0.4] - 1 Sep 2019

* Support of adding any StatelessWidget to VerticalMenu, It won't receive focus.

* VerticalMenuForAndroidTV now accepts list of StatelessWidget as menu items

* Support of FocusableSimpleMenuItem which can be added to VerticalMenuForAndroidTV

* Support of adding any widget to VerticalMenuForAndroidTV, it won't get focused

* Added example code in README.md

## [0.0.3] - 1 Sep 2019

* Bug fix, VerticalMenuItem needs to extend from StatelessWidget

## [0.0.2] - 31 Aug 2019

* VerticalMenuForAndroidTV and HyperLinkMenuItem supported,

* Container parameters are passed to VerticalMenuForAndroidTV which are used to display
the MenuItem

* Example updated

## [0.0.1] - 31 Aug 2019

* Initial Open Source release.

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MenuForAndroidTV());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/focusnode_widgets_test.dart

void main() {
  /*
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
   */
}


lib/focusnode_widgets.dart
library focusnode_widgets;

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class VerticalMenuForAndroidTV extends StatelessWidget {
  final List<Widget> menuItems;

  final Color nonFocusedBackgroundColor;
  final Color focusedBackgroundColor;
  final AlignmentGeometry alignment;
  final BoxConstraints constraints;
  final Decoration nonFocusedBackgroundDecoration;
  final Decoration focusedBackgroundDecoration;
  final EdgeInsetsGeometry padding;
  final Decoration nonFocusedForegroundDecoration;
  final Decoration focusedForegroundDecoration;
  final double width;
  final double height;
  final EdgeInsetsGeometry margin;
  final Matrix4 transform;

  const VerticalMenuForAndroidTV({
    Key key,
    @required this.menuItems,
    this.nonFocusedBackgroundColor,
    this.focusedBackgroundColor,
    this.alignment,
    this.constraints,
    this.nonFocusedBackgroundDecoration,
    this.focusedBackgroundDecoration,
    this.padding,
    this.nonFocusedForegroundDecoration,
    this.focusedForegroundDecoration,
    this.width,
    this.height,
    this.margin,
    this.transform,
  })  : assert(menuItems != null),
        assert(menuItems.length > 0),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    List<Widget> widgetList = List<Widget>();

    void constructMenuItem(Widget menuItem) {
      Widget widget;

      if (menuItem is _VerticalMenuItem) {
        _VerticalMenuItem verticalMenuItem = menuItem;

        widget = _FocusNodeEnterTapActionableWidget(
          child: verticalMenuItem.build(context),
          handleEnterTapAction: verticalMenuItem?.onEnterTapAction,
          autoFocus: verticalMenuItem?.autoFocus,
          debugLabel: verticalMenuItem?.debugLabel,
          voidTrapKeyEvent: verticalMenuItem?.voidTrapKeyEvent,
          nonFocusedForegroundDecoration: nonFocusedForegroundDecoration,
          transform: transform,
          nonFocusedBackgroundColor: nonFocusedBackgroundColor,
          focusedForegroundDecoration: focusedForegroundDecoration,
          nonFocusedBackgroundDecoration: nonFocusedBackgroundDecoration,
          height: height,
          focusedBackgroundDecoration: focusedBackgroundDecoration,
          focusedBackgroundColor: focusedBackgroundColor,
          constraints: constraints,
          width: width,
          margin: margin,
          padding: padding,
          alignment: alignment,
        );
      } else {
        widget = menuItem;
      }

      widgetList.add(widget);
    }

    menuItems.forEach(constructMenuItem);

    Column column = Column(
      children: widgetList,
    );

    Center center = Center(child: column);

    SingleChildScrollView singleChildScrollView = SingleChildScrollView(
      child: center,
    );

    return singleChildScrollView;
  }
}

abstract class _VerticalMenuItem extends StatelessWidget {
  final bool autoFocus;
  final String debugLabel;

  const _VerticalMenuItem({
    Key key,
    @required this.autoFocus,
    @required this.debugLabel,
    this.voidTrapKeyEvent,
  })  : assert(autoFocus != null),
        super();

  @protected
  void onEnterTapAction(BuildContext context);

  @protected
  final VoidTrapKeyEvent voidTrapKeyEvent;
}

/// Signature of callbacks that have BuildContext and return no data.
typedef VoidBuildContext = void Function(BuildContext context);

class FocusableSimpleMenuItem extends _VerticalMenuItem {
  final Widget child;
  final VoidBuildContext enterTapActionCallback;

  const FocusableSimpleMenuItem(
      {Key key,
      @required this.child,
      bool autoFocus = false,
      String debugLabel = "FocusableSimpleMenuItem",
      this.enterTapActionCallback})
      : assert(child != null),
        super(autoFocus: autoFocus, debugLabel: debugLabel);

  @protected
  const FocusableSimpleMenuItem.forPrintKey({
    Key key,
    @required this.child,
    bool autoFocus = false,
    String debugLabel = "FocusableSimpleMenuItem",
    this.enterTapActionCallback,
    @required VoidTrapKeyEvent voidTrapKeyEvent,
  })  : assert(child != null),
        super(
            autoFocus: autoFocus,
            debugLabel: debugLabel,
            voidTrapKeyEvent: voidTrapKeyEvent);

  @override
  void onEnterTapAction(BuildContext context) {
    if (enterTapActionCallback == null) {
      return;
    }
    enterTapActionCallback(context);
  }

  @override
  Widget build(BuildContext context) {
    return child;
  }
}

typedef VoidTrapKeyEvent = void Function(
    BuildContext context, RawKeyEvent event);

class KeyPrintMenuItem extends FocusableSimpleMenuItem {
  KeyPrintMenuItem({
    Key key,
    @required Widget child,
    bool autoFocus = false,
    String debugLabel = "KeyPrintMenuItem",
    @required VoidTrapKeyEvent voidTrapKeyEvent,
  }) : super.forPrintKey(
            autoFocus: autoFocus,
            debugLabel: debugLabel,
            child: child,
            voidTrapKeyEvent: voidTrapKeyEvent);
}

typedef VoidBuildContextValue = void Function(BuildContext context, bool value);

class CheckboxListTileMenuItem extends _VerticalMenuItem {
  final bool value;
  final VoidBuildContextValue enterTapActionCallback;
  final Color activeColor;
  final Widget title;
  final Widget subtitle;
  final Widget secondary;
  final bool isThreeLine;
  final bool dense;
  final bool selected;
  final ListTileControlAffinity controlAffinity;

  const CheckboxListTileMenuItem({
    Key key,
    @required this.value,
    @required this.enterTapActionCallback,
    this.activeColor,
    this.title,
    this.subtitle,
    this.isThreeLine = false,
    this.dense,
    this.secondary,
    this.selected = false,
    this.controlAffinity = ListTileControlAffinity.platform,
    bool autoFocus = false,
    String debugLabel = "CheckboxListTileMenuItem",
  })  : assert(value != null),
        assert(isThreeLine != null),
        assert(!isThreeLine || subtitle != null),
        assert(selected != null),
        assert(controlAffinity != null),
        super(key: key, autoFocus: autoFocus, debugLabel: debugLabel);

  @override
  void onEnterTapAction(BuildContext context) {
    print("onEnterTapAction");
    enterTapActionCallback(context, !value);
  }

  @override
  Widget build(BuildContext context) {
    void _onChanged(bool newValue) {
      enterTapActionCallback(context, newValue);
    }

    CheckboxListTile checkboxListTile = CheckboxListTile(
      key: key,
      value: value,
      onChanged: _onChanged,
      activeColor: activeColor,
      title: title,
      subtitle: subtitle,
      isThreeLine: isThreeLine,
      dense: dense,
      secondary: secondary,
      selected: selected,
      controlAffinity: controlAffinity,
    );
    return checkboxListTile;
  }
}

class HyperLinkMenuItem extends _VerticalMenuItem {
  final String displayText;
  final VoidBuildContext launchUrl;

  final TextStyle focusStyle = const TextStyle(
    fontWeight: FontWeight.normal,
    color: Colors.blue,
    decoration: TextDecoration.underline,
  );

  const HyperLinkMenuItem(
      {Key key,
      @required this.displayText,
      @required this.launchUrl,
      bool autoFocus = false,
      String debugLabel = "HyperLinkMenuItem",
      VoidCallback onEnterTapAction})
      : assert(displayText != null),
        assert(launchUrl != null),
        super(
          key: key,
          autoFocus: autoFocus,
          debugLabel: debugLabel,
        );

  @override
  void onEnterTapAction(BuildContext context) {
    if (launchUrl == null) {
      return;
    }
    launchUrl(context);
  }

  @override
  Widget build(BuildContext context) {
    Text underLinedText = Text(
      displayText,
      style: focusStyle,
    );

    return underLinedText;
  }
}

class _FocusNodeEnterTapActionableWidget extends StatefulWidget {
  final String debugLabel;
  final bool autoFocus;
  final Widget child;
  final VoidBuildContext _handleEnterTapAction;
  final VoidTrapKeyEvent voidTrapKeyEvent;

  final Color nonFocusedBackgroundColor;
  final Color focusedBackgroundColor;
  final AlignmentGeometry alignment;
  final BoxConstraints constraints;
  final Decoration nonFocusedBackgroundDecoration;
  final Decoration focusedBackgroundDecoration;
  final EdgeInsetsGeometry padding;
  final Decoration nonFocusedForegroundDecoration;
  final Decoration focusedForegroundDecoration;
  final double width;
  final double height;
  final EdgeInsetsGeometry margin;
  final Matrix4 transform;

  _FocusNodeEnterTapActionableWidget({
    Key key,
    VoidBuildContext handleEnterTapAction,
    this.debugLabel,
    @required this.autoFocus,
    @required this.child,
    @required this.nonFocusedBackgroundColor,
    @required this.focusedBackgroundColor,
    @required this.alignment,
    @required this.constraints,
    @required this.nonFocusedBackgroundDecoration,
    @required this.focusedBackgroundDecoration,
    @required this.padding,
    @required this.nonFocusedForegroundDecoration,
    @required this.focusedForegroundDecoration,
    @required this.width,
    @required this.height,
    @required this.margin,
    @required this.transform,
    this.voidTrapKeyEvent,
  })  : assert(child != null),
        _handleEnterTapAction = handleEnterTapAction,
        super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _FocusableEnterTapActionableWidget();
  }
}

class _FocusableEnterTapActionableWidget
    extends State<_FocusNodeEnterTapActionableWidget> {
  bool _gestureDetectorRequestedFocus = false;

  bool _handleOnKey(FocusNode node, RawKeyEvent event, BuildContext context) {
    if (widget.voidTrapKeyEvent != null) {
      widget.voidTrapKeyEvent(context, event);
      return true;
    }

    if (event is RawKeyDownEvent) {
      print(
          '_handleKeyPress: Focus node ${node.debugLabel} got key event: ${event.logicalKey}');
      if (event.logicalKey == LogicalKeyboardKey.arrowDown) {
        TraversalDirection direction = TraversalDirection.down;
        node.focusInDirection(direction);
        return true;
      } else if (event.logicalKey == LogicalKeyboardKey.arrowUp) {
        TraversalDirection direction = TraversalDirection.up;
        node.focusInDirection(direction);
        return true;
      } else if (/*event.logicalKey == LogicalKeyboardKey.enter ||*/
          event.logicalKey == LogicalKeyboardKey.select) {
        return _handleEnterTapAction(context);
      }
    }

    return false;
  }

  bool _handleEnterTapAction(BuildContext context) {
    if (widget._handleEnterTapAction == null) {
      return false;
    }
    widget._handleEnterTapAction(context);
    return true;
  }

  void _handleOnFocusChange(bool focusGained, BuildContext context) {
    if (focusGained) {
      print("focus gained by " + widget.debugLabel);
      if (_gestureDetectorRequestedFocus) {
        _gestureDetectorRequestedFocus = false;
        _handleEnterTapAction(context);
      }
      //TODO: need to show the child widget in focused state
    } else {
      //TODO: need to show the child widget in non focused state
      print("focus lossed by " + widget.debugLabel);
    }
  }

  Widget _getEnabledChild(bool hasFocus) {
    Container container = Container(
      child: widget.child,
      color: hasFocus
          ? widget.focusedBackgroundColor
          : widget.nonFocusedBackgroundColor,
      alignment: widget.alignment,
      constraints: widget.constraints,
      decoration: hasFocus
          ? widget.focusedBackgroundDecoration
          : widget.nonFocusedBackgroundDecoration,
      padding: widget.padding,
      foregroundDecoration: hasFocus
          ? widget.focusedForegroundDecoration
          : widget.nonFocusedForegroundDecoration,
      width: widget.width,
      height: widget.height,
      margin: widget.margin,
      transform: widget.transform,
    );

    return container;
  }

  @override
  Widget build(BuildContext context) {
    Builder builder = Builder(builder: (BuildContext context) {
      final FocusNode focusNode = Focus.of(context);
      final hasFocus = focusNode.hasFocus;

      GestureDetector gestureDetector = GestureDetector(
          onTap: () {
            if (!hasFocus) {
              _gestureDetectorRequestedFocus = true;
              focusNode.requestFocus();
            } else {
              _handleEnterTapAction(context);
            }
          },
          child: _getEnabledChild(hasFocus));

      return gestureDetector;
    });

    void _onFocusChange(bool focusGained) {
      _handleOnFocusChange(focusGained, context);
    }

    bool _onKey(FocusNode node, RawKeyEvent event) {
      return _handleOnKey(node, event, context);
    }

    Focus focusableEnterTapActionableChild = Focus(
      onFocusChange: _onFocusChange,
      autofocus: widget.autoFocus == null ? false : widget.autoFocus,
      onKey: _onKey,
      debugLabel: widget.debugLabel,
      child: builder,
    );

    return focusableEnterTapActionableChild;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:focusnode_widgets/focusnode_widgets.dart';

void main() => runApp(MenuForAndroidTV());

class MenuForAndroidTV extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _MenuForAndroidTV();
  }
}

class _MenuForAndroidTV extends State<MenuForAndroidTV> {
  final String _title = 'focusnode_widgets example for Android TV';
  final String _instructions =
      'Use DPAD up down arrow keyss to navigate and DPAD center key to open the URL';

  bool checked = false;

  @override
  Widget build(BuildContext context) {
    final TextTheme textTheme = Theme.of(context).textTheme;

    void launchUrl(BuildContext context) {
      // put code to launch the URL here
      // on android TV url needs to be launched in WebView as per guidelines,
      // browser can't be used
    }

    HyperLinkMenuItem flutterLink = HyperLinkMenuItem(
      displayText: 'Flutter - Beautiful native apps in record time',
      launchUrl: launchUrl,
      autoFocus: true,
    );

    void _onChanged(BuildContext context, bool selected) {
      checked = selected;
      setState(() {});
    }

    CheckboxListTileMenuItem checkboxListTileMenuItem =
        CheckboxListTileMenuItem(
      title: DefaultTextStyle(
          style: textTheme.headline,
          child: Text('I accept Terms and Conditions')),
      enterTapActionCallback: _onChanged,
      value: checked,
      controlAffinity: ListTileControlAffinity.leading,
    );

    Text nonFocusableText2 = Text('Non focusable text item 2');

    void _handleEnterTapAction(BuildContext context) {
      final snackBar = SnackBar(
        content: Text('Yay! A SnackBar!'),
        action: SnackBarAction(
          label: 'Undo',
          onPressed: () {
            // Some code to undo the change.
          },
        ),
      );
      Scaffold.of(context).showSnackBar(snackBar);
    }

    FocusableSimpleMenuItem focusableSimpleMenuItem = FocusableSimpleMenuItem(
      child: Text('Click, Tap or Press Enter/ OK to show SnackBar'),
      enterTapActionCallback: _handleEnterTapAction,
    );

    Text instructions = Text(_instructions);

    VerticalMenuForAndroidTV verticalMenuForAndroidTV =
        VerticalMenuForAndroidTV(
      menuItems: [
//        keyPrintMenuItem,
//        keyLogShort,
        flutterLink,
        nonFocusableText2,
        checkboxListTileMenuItem,
        focusableSimpleMenuItem,
        instructions,
      ],
      focusedBackgroundDecoration: BoxDecoration(
        border: Border.all(color: Colors.amber[900], width: 2),
      ),
      alignment: Alignment.center,
      constraints: BoxConstraints(
        maxHeight: 40,
        maxWidth: 700,
        minHeight: 10,
        minWidth: 100,
      ),
    );

    DefaultTextStyle menu = DefaultTextStyle(
      style: textTheme.headline,
      child: verticalMenuForAndroidTV,
    );

    return MaterialApp(
      title: _title,
      home: Scaffold(
        appBar: AppBar(title: Text(_title)),
        body: menu,
      ),
    );
  }
}
